src/Configuration/Programmer/Programmer_cmd.cpp

```diff
@@ -30,46 +36,108 @@ int main(int argc, const char** argv) {
   // programmer_entry(&cmdarg);
 
   // simple programmer API manual testing
+
+  std::string openOCDPath = "/home/guanyung/dev/openocd/src/openocd";
+  openOCDPath = "/home/asic01/development/openocd_rs/src/openocd";
+  int ret = InitLibrary(openOCDPath);
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+  }
   std::vector<Device> devices;
-  std::string outputMsg;
-  if (ListDevices(devices, outputMsg)) {
+
+  std::vector<FOEDAG::Cable> cables;
+  ret = GetAvailableCables(cables);
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+  }
+  CFG_POST_MSG("return %d", ret);
+  CFG_POST_MSG("-- GetAvailableCables(cables) --")
+  for (auto& cable : cables) {
+    CFG_POST_MSG("--------------")
+    CFG_POST_MSG("product ID: 0x%04x", cable.productId);
+    CFG_POST_MSG("vendor  ID: 0x%04x", cable.vendorId);
+    CFG_POST_MSG("serial    : %s", cable.serialNumber.c_str());
+    CFG_POST_MSG("desc      : %s", cable.description.c_str());
+    CFG_POST_MSG("bus       : %d", cable.busAddr);
+    CFG_POST_MSG("port      : %d", cable.portAddr);
+    CFG_POST_MSG("device    : %d", cable.deviceAddr);
+    CFG_POST_MSG("speed     : %d", cable.speed);
+    CFG_POST_MSG("transport : %d", cable.transport);
+    CFG_POST_MSG("channel   : %d", cable.channel);
+  }
+
+  CFG_POST_MSG("-- ListDevices(cables, outputMsg) --")
+  if (cables.size() > 0) {
+    ret = ListDevices(cables[0], devices);
+    if (ret != NoError) {
+      CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+      return ret;
+    }
     CFG_POST_MSG("### ListDevices API testing ###");
     for (auto& device : devices) {
       CFG_POST_MSG("-------------------------");
       CFG_POST_MSG("Device name: %s", device.name.c_str());
-      CFG_POST_MSG("Device index: %s", std::to_string(device.index).c_str());
-      CFG_POST_MSG("Device jtagId: %s", device.jtagId.c_str());
-      CFG_POST_MSG("Device mask: %s", std::to_string(device.mask).c_str());
-      CFG_POST_MSG("Device irlength: %s",
-                   std::to_string(device.irlength).c_str());
-      CFG_POST_MSG("Device flashSize: %s",
-                   std::to_string(device.flashSize).c_str());
+      CFG_POST_MSG("Device index: %d", device.index);
+      CFG_POST_MSG("Device jtagId: %d", device.tapInfo.idCode);
+      CFG_POST_MSG("Device mask: %d", device.tapInfo.irMask);
+      CFG_POST_MSG("Device irlength: %d", device.tapInfo.irLen);
+      CFG_POST_MSG("Device flashSize: %d", device.flashSize);
```

Comment:
Printing flash size in decimal is a mess as well. Should print it in unit (Will discuss more in weekly)